src/backend/distributed/executor/subplan_execution.c

```diff
@@ -81,3 +124,99 @@ ExecuteSubPlans(DistributedPlan *distributedPlan)
 		FreeExecutorState(estate);
 	}
 }
+
+
+/*
+ * RecordSubplanExecutionsOnNodes iterates over the usedSubPlanNodeList,
+ * and for each entry, record the workerNodes that are accessed by
+ * the distributed plan.
+ *
+ * Later, we'll use this information while we broadcast the intermediate
+ * results to the worker nodes. The idea is that the intermediate result
+ * should only be broadcasted to the worker nodes that are accessed by
+ * the distributedPlan(s) that the subPlan is used in.
+ *
+ * Finally, the function recursively descends into the actual subplans
+ * of the input distributedPlan as well.
+ */
+static void
+RecordSubplanExecutionsOnNodes(HTAB *intermediateResultsHash,
+							   DistributedPlan *distributedPlan)
+{
+	List *usedSubPlanNodeList = distributedPlan->usedSubPlanNodeList;
+	ListCell *usedSubPlanNodeCell = NULL;
+	List *subPlanList = distributedPlan->subPlanList;
+	ListCell *subPlanCell = NULL;
+
+	foreach(usedSubPlanNodeCell, usedSubPlanNodeList)
+	{
+		Const *resultIdConst = lfirst(usedSubPlanNodeCell);
+		Datum resultIdDatum = resultIdConst->constvalue;
+		char *resultId = TextDatumGetCString(resultIdDatum);
+
+		IntermediateResultHashKey key;
+		IntermediateResultHashEntry *entry = NULL;
+		bool found = false;
+
+		strlcpy(key.intermediate_result_id, resultId, NAMEDATALEN);
+		entry = hash_search(intermediateResultsHash, &key, HASH_ENTER, &found);
+
+		if (!found)
+		{
+			entry->nodeList = NIL;
+		}
+
+		if (list_length(entry->nodeList) == GetWorkerNodeCount())
+		{
+			entry->containsAllNodes = true;
```

Comment:
We should just break here, also when this condition occurs in AppendAllAccessedWorkerNodes.

Reply:
makes sense. Removed the unused flag, and added early returns in 2 places in commit 5df2a6c471421e9c9a5a4d526860bae9d89a7968